### PR TITLE
buffer redesign - 2.) memory implementations, basic ctypes

### PR DESCRIPTION
**Pull request by jerch**
The PR provides the basic memory implementations based on typed arrays "heaps" and basic c types to be used with the memory.

memory allocator types:
- `StackMemory` allocator like stack in C
- `PoolMemory` allocator for fixed block sizes
- `SeglistMemory` general purpose allocator like malloc/free in C

c types:
- signed/unsigned number types up to 32 bit
- pointer types
- array types
- struct types

This PR does not add any functionality to xterm,js, it just provides the basic data structures to be used later on with upcoming buffer changes.
Up for a review.

cc: @Tyriar @mofux @bgw

### CODE REVIEW

**Comment by jerch** [comment]
Where to put `Memory.ts`? It will be part of the smelly dark corner deep in the basement hidden behind several storage classes to come. Most likely it will not be used directly by `InputHandler` and such. Frontend stuff will not need it at all. So where to put?

Up for reviews :smiley_cat:

File: src/Memory.test.ts [hunk 1]
```diff
@@ -0,0 +1,591 @@
...
+describe('Memory', function (): void {
+  describe('StackMemory', function (): void {
+    let mem: StackMemory;
+    beforeEach(function (): void {
+      mem = new StackMemory(8);
+    });
+    it('init', function (): void {
+      chai.expect(mem.sp).equals(mem.RESERVED_BYTES >>> 2);
+      chai.expect(mem.data.length).equals((8 + mem.RESERVED_BYTES) >>> 2);
+    });
```
Comment (Tyriar) [inline src/Memory.test.ts]: as

File: src/Memory.test.ts [hunk 2]
```diff
@@ -0,0 +1,591 @@
...
+import * as chai from 'chai';
+import { StackMemory, PoolMemory, IMemory, AccessType, AccessBits, SeglistMemory, SL, Address, ctypes } from './Memory';
+
+describe('Memory', function (): void {
+  describe('StackMemory', function (): void {
+    let mem: StackMemory;
+    beforeEach(function (): void {
+      mem = new StackMemory(8);
+    });
+    it('init', function (): void {
```
Comment (Tyriar) [inline src/Memory.test.ts]: as

**Comment by jerch** [comment]
Closing this PR as it seems it will not be needed.

**Pull request closed without merging**